Add the execution support for segmented aggregation

zacw7 opened the pull request:
Depends on #17458 

Benchmark:
```
Benchmark                                                (operatorType)  (rowsPerSegment)  Mode  Cnt    Score   Error  Units
BenchmarkHashAndSegmentedAggregationOperators.benchmark       segmented                 1  avgt   30  109.391 ± 7.813  ms/op
BenchmarkHashAndSegmentedAggregationOperators.benchmark       segmented                10  avgt   30   52.914 ± 5.455  ms/op
BenchmarkHashAndSegmentedAggregationOperators.benchmark       segmented               800  avgt   30   28.937 ± 5.291  ms/op
BenchmarkHashAndSegmentedAggregationOperators.benchmark       segmented            100000  avgt   30    6.492 ± 0.184  ms/op
BenchmarkHashAndSegmentedAggregationOperators.benchmark            hash                 1  avgt   30   18.439 ± 1.193  ms/op
BenchmarkHashAndSegmentedAggregationOperators.benchmark            hash                10  avgt   30   18.707 ± 2.586  ms/op
BenchmarkHashAndSegmentedAggregationOperators.benchmark            hash               800  avgt   30   17.132 ± 0.495  ms/op
BenchmarkHashAndSegmentedAggregationOperators.benchmark            hash            100000  avgt   30    5.660 ± 0.174  ms/op
```
Manual testing(717,977,748,003 rows / 3.02 TB):
|                               | QueryID                     | Splits | Latency  | CPU         | Memory    | Per wall sec |
|-------------------------------|-----------------------------|--------|----------|-------------|-----------|--------------|
|                      Baseline | 20220506_014308_00006_iiy5d |  84.9K |  36.96 s | 23.54 hours | 416.32 GB | 83.51 GB     |
|      File Splittable Disabled | 20220506_014507_00008_iiy5d |  6.86K | 1.50 min | 20.49 hours | 405.88 GB | 34.32 GB     |
| Segmented Aggregation Enabled | 20220506_015126_00011_iiy5d |  2.06K | 1.88 min | 33.31 hours | 30.60 GB  | 27.37 GB     |

Latency regression is observed during testing, which is expected. In order to enable segmented aggregation, splitting files needs to be disabled to preserve the order. As the result, much less splits are generated and it decreased the table scan concurrency drastically especially when there are a lot of big files to scan.


```
== RELEASE NOTES ==

General Changes
* Add ability to flush the aggregated data when the current input segment is exhausted. It reduces the memory footprint and improves the performance of aggregation when the data is already ordered by a subset of the group-by keys.
This can be enabled with the ``segmented_aggregation_enabled`` session property or the ``optimizer.segmented-aggregation-enabled`` configuration property.

Hive Changes
* Add support for segmented aggregation to reduce the memory footprint and improve query performance when the order-by keys are a subset of the group-by keys. This can be enabled with the ``order_based_execution_enabled`` session property or the ``hive.order-based-execution-enabled`` configuration property.
```

kewang1024 reviewed (commented):
Haven't finished reviewing the segmented aggregation logic inside `HashAggregationOperator`, will first post some of the suggestions here
1. We need some changes to release note to Hive Changes:
there is no segmented aggregation logic added to hive connector, for hive connector, we added the ``order_based_execution_enabled`` session property and the ``hive.order-based-execution-enabled`` configuration property to disable splitting and expose the data property to leverage segmented aggregation
2.  For refactor commit, one optional suggestion is to add comments in the class to explain the functions of different fields, and a succinct summary of logic within key functions (needsInput, addInput, getOutput, finish etc). I think it would help people understand better when you introduce extra logic for segmented aggregation in the later commit

zacw7 commented on presto-main/src/main/java/com/facebook/presto/sql/planner/LocalExecutionPlanner.java:
```diff
@@ -3154,7 +3159,7 @@ private OperatorFactory createHashAggregationOperatorFactory(
                     .collect(toImmutableList());
 
-            if (isStreamable) {
+            if (isStreamable && preGroupbyVariables.size() == groupbyVariables.size()) {
```
The check is not for segmented aggregation but for streaming aggregation right?

kewang1024 replied:
You can take a look at the definition of `isStreamable`, it already includes the logic of checking the size

Or maybe you need to pull the newest code if you can't see it in your local right now

zacw7 replied:
Got it. Thanks for the pointer!

zacw7 commented on presto-main/src/test/java/com/facebook/presto/operator/TestHashAggregationOperator.java:
```diff
@@ -702,6 +727,51 @@ public void testMask()
         assertEquals(outputPage.getBlock(1).getLong(0), 1L);
     }
 
+    @Test
```
[StreamingAggregationOperator.java](https://github.com/prestodb/presto/blob/a4dea8ced0430b8b933f79b5bb2bd14ddaeb9738/presto-main/src/main/java/com/facebook/presto/operator/StreamingAggregationOperator.java) is a separate operator while segmented aggregation is part of HashAggregationOperator.

kewang1024 replied:
Yeah, the reason why I suggested moving to a new class is that you would have many different cases specific to segmented aggregation that we should test, and moving it to a new class would be clearer IMO

zacw7 commented on presto-main/src/test/java/com/facebook/presto/operator/BenchmarkHashAndStreamingAggregationOperators.java:
```diff
@@ -127,7 +128,12 @@ public void setup()
 
             if (hashAggregation) {
-                operatorFactory = createHashAggregationOperatorFactory(pagesBuilder.getHashChannel());
+                if (operatorType.equalsIgnoreCase("segmented")) {
+                    operatorFactory = createHashAggregationOperatorFactory(pagesBuilder.getHashChannel(), ImmutableList.of(VARCHAR, BIGINT), ImmutableList.of(0));
+                }
+                else {
+                    operatorFactory = createHashAggregationOperatorFactory(pagesBuilder.getHashChannel(), ImmutableList.of(), ImmutableList.of());
+                }
```
IIUC, segmented aggregation essentially is still hash aggregation - just that in some cases we can flush the output when possible and rebuild the hash. The proposed code structure seems like they are mutually exclusive. wdyt?

kewang1024 replied:
Mine is the same logic as yours, just less nested logic, it's the same as
if streaming aggregation
else if hash aggregation
        if segmented hash aggregation
        if normal hash aggregation

zacw7 commented on presto-main/src/main/java/com/facebook/presto/operator/HashAggregationOperator.java:
```diff
@@ -54,8 +58,10 @@
     {
         private final int operatorId;
         private final PlanNodeId planNodeId;
+        private final List<Type> sourceTypes;
```
Great catch!

zacw7 commented on presto-main/src/test/java/com/facebook/presto/operator/OperatorAssertion.java:
```diff
@@ -114,7 +114,7 @@ public static List<Page> finishOperator(Operator operator)
         ImmutableList.Builder<Page> outputPages = ImmutableList.builder();
 
-        for (int loopsSinceLastPage = 0; !operator.isFinished() && loopsSinceLastPage < 1_000; loopsSinceLastPage++) {
+        for (int loopsSinceLastPage = 0; !operator.isFinished() && loopsSinceLastPage < 100_000; loopsSinceLastPage++) {
```
Seems not needed anymore.

zacw7 commented on presto-main/src/test/java/com/facebook/presto/operator/BenchmarkHashAndStreamingAggregationOperators.java:
```diff
@@ -127,7 +128,12 @@ public void setup()
             if (hashAggregation) {
-                operatorFactory = createHashAggregationOperatorFactory(pagesBuilder.getHashChannel());
+                if (operatorType.equalsIgnoreCase("segmented")) {
+                    operatorFactory = createHashAggregationOperatorFactory(pagesBuilder.getHashChannel(), ImmutableList.of(VARCHAR, BIGINT), ImmutableList.of(0));
```
Makes sense. Created one comparing hash and segmented aggregation.

zacw7 commented:
Hey @kewang1024, sorry for the delay. I've addressed most of the comments including creating a new benchmark for hash vs segmented aggregation. For the test for functionality and correctness tests, if a separate file is needed, could you please elaborate more about what other cases should be covered?

kewang1024 commented:
> Hey @kewang1024, sorry for the delay. I've addressed most of the comments including creating a new benchmark for hash vs segmented aggregation. For the test for functionality and correctness tests, if a separate file is needed, could you please elaborate more about what other cases should be covered?

What I have in mind initially is to cover the cases in [Streaming Aggregation test](https://github.com/mbasmanova/presto/blob/cc916c4af244cd04d2ef14c3ffad2942a656b431/presto-main/src/test/java/com/facebook/presto/operator/TestStreamingAggregationOperator.java)

kewang1024 reviewed (commented):
For SegmentedAggregation benchmark, the performance benchmark result shows segmented aggregation is much worse than hash aggregation, we need to have an investigation on what’s causing the performance downgrading

kewang1024 commented on presto-main/src/main/java/com/facebook/presto/operator/HashAggregationOperator.java:
```diff
@@ -349,58 +350,25 @@ public boolean needsInput()
-            return unfinishedWork == null;
+            return unfinishedWork == null && currentPage == null;
         }
     }
```
NIT let's also include this change to make this function easier to understand
```
    // This operator only needs input when
    // - 1. It hasn't received finish() signal (more input to come).
    // - 2. Page has been processed.
    // - 3. Aggregation has been processed.
    // - 4. If this is partial aggregation and it hasn't reached memory limit.
    @Override
    public boolean needsInput()
    {
        return !finishing &&
                unfinishedWork == null &&
                outputPages == null &&
                partialAggregationReachedMemoryLimit();
    }
```

kewang1024 replied:
NIT let's also include this change to make this function easier to understand
```
    // This operator only needs input when
    // - 1. It hasn't received finish() signal (more input to come).
    // - 2. Page has been processed.
    // - 3. Aggregation has been processed.
    // - 4. If this is partial aggregation and it hasn't reached memory limit.
    @Override
    public boolean needsInput()
    {
        return !finishing &&
                unfinishedWork == null &&
                outputPages == null &&
                partialAggregationReachedMemoryLimit();
    }
```

kewang1024 commented on presto-main/src/main/java/com/facebook/presto/operator/HashAggregationOperator.java:
```diff
@@ -349,58 +350,25 @@ public boolean needsInput()
-        }
+        initializeAggregationBuilder();
 
         // process the current page; save the unfinished work if we are waiting for memory
```
we can extract this logic to a common function as well
```
    // Process the current page; save the unfinished work if we are waiting for memory
    // Return true if page has been processed, else false
    private boolean processCurrentPage()
    {
        if (unfinishedWork == null)
            return true;

        boolean workDone = unfinishedWork.process();
        aggregationBuilder.updateMemory();
        if (workDone)
            unfinishedWork = null;

        return workDone;
    }
```

kewang1024 reviewed (commented):
1. For the first PR:
Let's start a new PR for refactor so that we can quickly merge it, I have left all the change suggestion in comments
 
2. For the second PR: The design where we flush by segment 
1) It's very costly  performance wise, because it closes and rebuilds hashBuilder for each segment  especially when one page has multiple segment
2) It makes the logic too complicated
We're introducing too many variables which makes the `HashAggregationOperator` logic more error-prone

My suggestion
1. We set a threshold for segmented aggregation flushing
2. We process by the minimum unit of Page (same as the current `HashAggregationOperator` behavior) 
3. It will trigger the flush when threshold limit is hit
